kolibri/core/assets/src/api_resource.js

```diff
@@ -105,46 +251,71 @@ class Collection {
-          // paginated data! Check to see if it is paginated.
-          if (typeof response.entity.results !== 'undefined') {
-            // Paginated objects have 'results' as their results object so interpret this as
-            // such.
-            this.set(response.entity.results);
-            this.pageCount = Math.ceil(response.entity.count / this.pageSize);
-          } else {
-            // It's all gone a bit Pete Tong.
-            logging.debug('Data appears to be malformed', response.entity);
-          }
+          this.synced = false;
+          client({ path: this.url, params }).then((response) => {
+            // Reset current models to only include ones from this fetch.
+            this.models = [];
+            this._model_map = {};
+            // Set response object - an Array - on the Collection to record the data.
+            // First check that the response *is* an Array
+            if (Array.isArray(response.entity)) {
+              this.set(response.entity);
+            } else {
+              // If it's not, there are two possibilities - something is awry, or we have received
+              // paginated data! Check to see if it is paginated.
+              if (typeof response.entity.results !== 'undefined') {
+                // Paginated objects have 'results' as their results object so interpret this as
+                // such.
+                this.set(response.entity.results);
+                this.pageCount = Math.ceil(response.entity.count / this.pageSize);
+              } else {
+                // It's all gone a bit Pete Tong.
+                logging.debug('Data appears to be malformed', response.entity);
+              }
+            }
+            // Mark that the fetch has completed.
+            this.synced = true;
+            this.models.forEach((model) => {
+              model.synced = true; // eslint-disable-line no-param-reassign
+            });
+            // Return the data from the models, not the models themselves.
+            resolve(this.data);
+            // Clean up the reference to this promise
+            this.promises.splice(this.promises.indexOf(promise), 1);
+          }, (response) => {
+            logging.error('An error occurred', response);
+            reject(response);
+            // Clean up the reference to this promise
+            this.promises.splice(this.promises.indexOf(promise), 1);
+          });
         }
-        // Mark that the fetch has completed.
-        this.synced = true;
-        this.models.forEach((model) => {
-          model.synced = true; // eslint-disable-line no-param-reassign
-        });
-        // Return the data from the models, not the models themselves.
-        resolve(this.data);
+      },
+      (reason) => {
+        reject(reason);
+      });
+    });
+    this.promises.push(promise);
+    return promise;
+  }
+
+  getCurrentFacility() {
```

Comment:
The neater solution might be to put this method on the facility resource
definition.

On Sun, 31 Jul 2016, 00:27 Eli, <notifications@github.com> wrote:

> In kolibri/core/assets/src/api_resource.js
> <https://github.com/learningequality/kolibri/pull/313#discussion_r72899717>
> :
>
> > -        this.synced = true;
> > -        this.models.forEach((model) => {
> > -          model.synced = true; // eslint-disable-line no-param-reassign
> > -        });
> > -        // Return the data from the models, not the models themselves.
> > -        resolve(this.data);
> > +      },
> > +      (reason) => {
> > +        reject(reason);
> > +      });
> > +    });
> > +    this.promises.push(promise);
> > +    return promise;
> > +  }
> > +
> > +  getCurrentFacility() {
>
> Yeah, I agree, bootstrapping the facility id into html is better. But be
> able to fetch this info from a endpoint may comes handy sometimes say when
> change facility without a page reload. And it is an awkward to place this
> function here. Since we already have a helper function getCookie in
> api_resuorce, would it make sense to have the core app exposing a utility
> suit?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/learningequality/kolibri/pull/313/files/117c766d44e30be8171530f129e53486f9923b8e#r72899717>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABmkvTSo1ALmlWYA-D8KZDNhx7mi07gUks5qbE5SgaJpZM4JXxCB>
> .
>